cmd/release/runner.go

```diff
@@ -0,0 +1,249 @@
+	err := r.replaceVersionInFile(file, r.flag.WorkInProgressVersion, r.flag.CurrentVersion)
+	if err != nil {
+		return plumbing.Hash{}, microerror.Mask(err)
+	}
+
+	commit, err := r.addAndCommitChanges(file, worktree, r.flag.Author, fmt.Sprintf("release %s", r.flag.TagName))
+	if err != nil {
+		return plumbing.Hash{}, microerror.Mask(err)
+	}
+
+	return commit, err
+}
+
+// replaceWorkInProgressVersionWithRelease replaces work in progress version with the release that we want to publish in the source code.
+func (r *runner) replaceReleaseVersionWithNextWorkInProgress(file string, worktree *git.Worktree) (plumbing.Hash, error) {
+	err := r.replaceVersionInFile(file, r.flag.CurrentVersion, r.flag.NextPatchWorkInProgressVersion)
+	if err != nil {
+		return plumbing.Hash{}, microerror.Mask(err)
+	}
+
+	commit, err := r.addAndCommitChanges(file, worktree, r.flag.Author, fmt.Sprintf("bump new work in progress version to %s", r.flag.NextPatchWorkInProgressVersion))
+	if err != nil {
+		return plumbing.Hash{}, microerror.Mask(err)
+	}
+
+	return commit, err
+}
+
+func (r *runner) addAndCommitChanges(file string, worktree *git.Worktree, author *object.Signature, commitMessage string) (plumbing.Hash, error) {
+	_, err := worktree.Add(strings.TrimPrefix(file, r.flag.RepositoryPath+"/"))
+	if err != nil {
+		return plumbing.Hash{}, microerror.Mask(err)
+	}
+
+	commit, err := worktree.Commit(commitMessage, &git.CommitOptions{
+		Author: author,
+	})
+
+	return commit, microerror.Mask(err)
+}
+
+func (r *runner) replaceVersionInFile(file, search, replaceWith string) error {
+	f, err := ioutil.ReadFile(file)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	filecontents := string(f)
+
+	if !strings.Contains(filecontents, search) {
+		return microerror.Maskf(executionFailedError, "No version was found in %s", file)
+	}
+
+	updatedFileContents := []byte(strings.Replace(filecontents, search, replaceWith, 1))
+	err = ioutil.WriteFile(file, updatedFileContents, 0)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *runner) addReleaseToChangelog(file string, worktree *git.Worktree) (plumbing.Hash, error) {
+	// Add new entry to changelog
+	search := "## [Unreleased]"
+	replaceWith := fmt.Sprintf("## [Unreleased]\n\n## [%s] %s", r.flag.CurrentVersion, time.Now().Format("2006-01-02"))
+	f, err := ioutil.ReadFile(file)
+	if err != nil {
+		return plumbing.Hash{}, microerror.Mask(err)
+	}
+	filecontents := string(f)
+
+	if !strings.Contains(filecontents, search) {
+		return plumbing.Hash{}, microerror.Maskf(executionFailedError, "No unreleased work was found in %s", file)
+	}
+
+	updatedFileContents := []byte(strings.Replace(filecontents, search, replaceWith, 1))
+	err = ioutil.WriteFile(file, updatedFileContents, 0)
+	if err != nil {
+		return plumbing.Hash{}, microerror.Mask(err)
+	}
+
+	// Update links at the bottom
+	m1 := regexp.MustCompile(`^(\[Unreleased]:)(.*)(v[0-9]+\.[0-9]+\.[0-9]+)(...HEAD)\n`)
```

Comment:
There is also a corner case of the first release https://raw.githubusercontent.com/giantswarm/template-operator/master/CHANGELOG.md. But we can iterate on that.